atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CassandraRepairHelper.java

```diff
@@ -16,67 +16,115 @@
 
     public CassandraRepairHelper(
             Function<Namespace, CassandraKeyValueServiceConfig> keyValueServiceConfigFactory,
             Function<Namespace, KeyValueService> keyValueServiceFactory) {
         this.keyValueServiceConfigFactory = keyValueServiceConfigFactory;
         this.keyValueServiceFactory = keyValueServiceFactory;
+
+        this.cqlClusters = Caffeine.newBuilder()
+                .maximumSize(100)
+                .expireAfterAccess(Duration.ofMinutes(10L))
+                .build(this::getCqlClusterUncached);
+    }
+
+    private CqlCluster getCqlClusterUncached(Namespace namespace) {
+        CassandraKeyValueServiceConfig config = keyValueServiceConfigFactory.apply(namespace);
+        return CqlCluster.create(config);
     }
 
-    public void repairInternalTables(
-            Namespace namespace, Consumer<Map<InetSocketAddress, Set<Range<LightweightOppToken>>>> repairTable) {
+    public void repairInternalTables(Namespace namespace, BiConsumer<String, RangesForRepair> repairTable) {
         KeyValueService kvs = keyValueServiceFactory.apply(namespace);
-        CqlCluster cqlCluster = getCqlCluster(namespace);
-        kvs.getAllTableNames().stream()
-                .filter(TABLES_TO_REPAIR::contains)
-                .map(TableReference::getTableName)
+        CqlCluster cqlCluster = cqlClusters.get(namespace);
+        KeyedStream.of(getTableNamesToRepair(kvs))
                 .map(tableName -> getRangesToRepair(cqlCluster, namespace, tableName))
                 // TODO(gs): this will do repairs serially, instead of batched. Is this fine? Port batching from
                 //   internal product?
                 .forEach(repairTable);
     }
 
-    private CqlCluster getCqlCluster(Namespace namespace) {
-        CassandraKeyValueServiceConfig config = keyValueServiceConfigFactory.apply(namespace);
-        return CqlCluster.create(config);
+    private Stream<String> getTableNamesToRepair(KeyValueService kvs) {
+        return kvs.getAllTableNames().stream()
+                .filter(TABLES_TO_REPAIR::contains)
+                .map(TableReference::getTableName);
+    }
+
+    public void repairTransactionsTables(
+            Namespace namespace,
+            Map<FullyBoundedTimestampRange, Integer> coordinationMap,
+            BiConsumer<String, RangesForRepair> repairTable) {
+        List<TransactionsTableInteraction> transactionsTableInteractions =
+                TransactionsTableInteraction.getTransactionTableInteractions(
+                        coordinationMap, DefaultRetryPolicy.INSTANCE);
+
+        Map<String, RangesForRepair> tokenRangesForRepair =
+                getRangesForRepairByTable(namespace, transactionsTableInteractions);
+
+        KeyedStream.stream(tokenRangesForRepair).forEach((table, ranges) -> {
+            log.info("Repairing ranges for table", SafeArg.of("table", table));
+            repairTable.accept(table, ranges);
+        });
```

Comment:
perhaps a little overkill for what a simple `entrySet().stream()` can achieve, but I respect the usage of `KeyedStream`.

Reply:
yeah, we actually just need `tokenRangesForRepair.forEach` here.